File: lclang/src/main/java/lclang/libs/lang/classes/numbers/IntClass.java
package lclang.libs.lang.classes.numbers;

import lclang.Global;
import lclang.libs.lang.classes.StringClass;
import lclang.types.Type;
import lclang.types.Types;

public class IntClass extends NumberClass {
    public static final String NAME = "int";
    public static final IntClass INSTANCE = new IntClass();
    public static final Type TYPE = INSTANCE.classType;
    public int value;

    private IntClass() {
        super(NAME);
        constructor = method((caller, lcClasses) -> IntClass.get(
                Integer.parseInt(lcClasses.get(0).toString(caller))), Types.ANY, TYPE);
    }

    private IntClass(int value) {
        super(NAME, value);
        this.value = value;
        globals.put("toString", method((caller, lcClasses) -> StringClass.get(String.valueOf(value)),
                StringClass.type));
        globals.put("toBinary", method((caller, lcClasses) -> StringClass.get(Integer.toBinaryString(value)),
                StringClass.type));
    }

    public static IntClass get(int value) {
        if(Global.cashedClasses.containsKey(value))
            return (IntClass) Global.cashedClasses.get(value);

        IntClass intClass = new IntClass(value);
        Global.saveCache(value, intClass);
        return intClass;
    }
}

File: lclang/src/main/kotlin/lclang/libs/lang/IntClass.kt
package lclang.libs.lang

import lclang.Caller
import lclang.constructor
import lclang.types.Types

class IntClass: NumberClass {
    var int: Int = 0
    private constructor(): super("int")
    private constructor(int: Int): super("int", int) {
        this.int = int
    }

    init {
        constructor = constructor(listOf(Types.ANY)) { list ->
            get(list[0].toString(this).toInt())
        }
    }

    override fun equals(other: Any?): Boolean {
        if(other is IntClass)
            return other.int == int

        return false
    }

    override fun toString(caller: Caller): String = int.toString()
    override fun hashCode(): Int = int.hashCode()

    companion object {
        val instance = IntClass()
        val integers = HashMap<Int, IntClass>()

        fun get(int: Int): IntClass {
            if(integers.containsKey(int))
                return integers[int]!!

            val intClazz = IntClass(int)
            integers[int] = intClazz
            return intClazz
        }
    }
}